firedrake/assemble.py

```diff
@@ -381,167 +399,164 @@ def vec(testmap, i):
+    for indices, kinfo in kernels:
+        kernel = kinfo.kernel
+        integral_type = kinfo.integral_type
+        domain_number = kinfo.domain_number
+        subdomain_id = kinfo.subdomain_id
+        coeff_map = kinfo.coefficient_map
+        pass_layer_arg = kinfo.pass_layer_arg
+        needs_orientations = kinfo.oriented
+        needs_cell_facets = kinfo.needs_cell_facets
+        needs_cell_sizes = kinfo.needs_cell_sizes
+
+        m = domains[domain_number]
+        subdomain_data = f.subdomain_data()[m]
+        # Find argument space indices
+        if is_mat:
+            i, j = indices
+        elif is_vec:
+            i, = indices
         else:
-            zero_tensor()
-        for indices, kinfo in kernels:
-            kernel = kinfo.kernel
-            integral_type = kinfo.integral_type
-            domain_number = kinfo.domain_number
-            subdomain_id = kinfo.subdomain_id
-            coeff_map = kinfo.coefficient_map
-            pass_layer_arg = kinfo.pass_layer_arg
-            needs_orientations = kinfo.oriented
-            needs_cell_facets = kinfo.needs_cell_facets
-            needs_cell_sizes = kinfo.needs_cell_sizes
-
-            m = domains[domain_number]
-            subdomain_data = f.subdomain_data()[m]
-            # Find argument space indices
-            if is_mat:
-                i, j = indices
-            elif is_vec:
-                i, = indices
-            else:
-                assert len(indices) == 0
-
-            sdata = subdomain_data.get(integral_type, None)
-            if integral_type != 'cell' and sdata is not None:
-                raise NotImplementedError("subdomain_data only supported with cell integrals.")
-
-            # Extract block from tensor and test/trial spaces
-            # FIXME Ugly variable renaming required because functions are not
-            # lexical closures in Python and we're writing to these variables
-            if is_mat and result_matrix.block_shape > (1, 1):
-                tsbc = []
-                trbc = []
-                # Unwind ComponentFunctionSpace to check for matching BCs
-                for bc in bcs:
-                    fs = bc.function_space()
-                    if fs.component is not None:
-                        fs = fs.parent
-                    if fs.index == i:
+            assert len(indices) == 0
+
+        sdata = subdomain_data.get(integral_type, None)
+        if integral_type != 'cell' and sdata is not None:
+            raise NotImplementedError("subdomain_data only supported with cell integrals.")
+
+        # Extract block from tensor and test/trial spaces
+        # FIXME Ugly variable renaming required because functions are not
+        # lexical closures in Python and we're writing to these variables
+        tsbc = []
+        trbc = []
+        if is_mat and result_matrix.block_shape > (1, 1):
+            # Unwind ComponentFunctionSpace to check for matching BCs
+            def collect_dbc(bcs):
```

Comment:
`collect_dbc` is defined twice. Can we try to merge the definition somehow and define it once? Perhaps if `collect_dbc` takes two arguments: `result_matrix`, and `bc`. Then the check for `block_shape` can be added to the definition of the function itself:
```python
def collect_dbc(result_matrix, bcs):
    if bcs is not None:
        if result_matrix.block_shape > (1, 1):
            for bc in bcs:
                fs = bc.function_space()
                if fs.component is not None:
                    fs = fs.parent
                if fs.index == i:
                    tsbc.append(bc)
                if fs.index == j:
                    if isinstance(bc, DirichletBC):
                        trbc.append(bc)
                    elif isinstance(bc, EquationBCSplit):
                        collect_dbc(bc.bcs)
                    else:
                        raise NotImplementedError("Unsupported bc type")
        else:
            if bcs is not None:
                for bc in bcs:
                    tsbc.append(bcs)
                    if isinstance(bc, DirichletBC):
                        trbc.append(bc)
                    elif isinstance(bc, EquationBCSplit):
                        collect_dbc(bc.bcs)

if is_mat:
    collect_dbc(bcs)
```

Might be a better way to refactor this, but just a suggestion.

Reply:
Fixed.